Don't rely on testing methods to be called in mocha context

## Informations

Type | Version
---|---
Improvement | 4.32.2

****

## Description
Right now, the testing helpers rely on being called from a mocha (`beforeEach`, `it`, etc) context. A 'static' injector is created here:

https://github.com/Romakita/ts-express-decorators/blob/07d73f52ade047c4b3f36b7e2ed6c0440d784faa/packages/testing/src/bootstrap.ts#L22

and used here:

https://github.com/Romakita/ts-express-decorators/blob/07d73f52ade047c4b3f36b7e2ed6c0440d784faa/packages/testing/src/inject.ts#L19

My goal is to be able to run the testing helpers from any context. Like this:

```ts
function bootstrapAsPromised(ServerConstructor: any): Promise<any> {
    return new Promise((res) => {
        bootstrap(ServerConstructor)(res);
    });
}
beforeEach(() => {
   // Do stuff...
  await bootstrapAsPromised();
});
```

Right now, this results in this error:

```
 TypeError: Cannot set property '$$injector' of undefined
      at before (C:\z\github\stryker-mutator\stryker-dashboard\packages\website-backend\node_modules\@tsed\testing\src\bootstrap.ts:22:20)
```

This is unexpected and unnecessary IMHO. We shouldn't rely on `this` pointer pollution here. I propose creating a file called `staticInjector.ts` and use that to get/set the injector in a singleton or something.

Ok, I'm trying to get things going. How do you build locally? The contributing.md file seems out-of-date. `npm install` delivers a ton of babel warnings: 

```
pm WARN @babel/preset-env@7.0.0-beta.47 requires a peer of @babel/core@7.0.0-beta.47 but none is installed. You must install peer dependencies yourself.
npm WARN @babel/plugin-proposal-async-generator-functions@7.0.0-beta.47 requires a peer of @babel/core@7.0.0-beta.47 but none is installed. You must install peer dependencies yourself.
npm WARN @babel/plugin-proposal-optional-catch-binding@7.0.0-beta.47 requires a peer of @babel/core@7.0.0-beta.47 but none is installed. You must install peer dependencies yourself.
npm WARN @babel/plugin-syntax-async-generators@7.0.0-beta.47 requires a peer of @babel/core@7.0.0-beta.47 but none is installed. You must install peer dependencies yourself.
...
```

Ignoring those and running `npm t` results in a lot of these errors:  

```
test/units/di/services/InjectorService.spec.ts:224:73 - error TS2345: Argument of type 'undefined' is not assignable to parameter of type '{}'.

224         this.getStub = Sinon.stub(this.injector, "getProvider").returns(undefined);
                                                                            ~~~~~~~~~

test/units/di/services/InjectorService.spec.ts:263:99 - error TS2345: Argument of type '{ buildable: true; injectable: true; }' is not assignable to parameter of type 'RegistrySettings'.
  Property 'registry' is missing in type '{ buildable: true; injectable: true; }' but required in type 'RegistrySettings'.

263         this.getRegistrySettingsStub = Sinon.stub(GlobalProviders, "getRegistrySettings").returns({
                                                                                                      ~
264           buildable: true,
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
265           injectable: true
    ~~~~~~~~~~~~~~~~~~~~~~~~~~
266         });
    ~~~~~~~~~
```

I can try to create this feature without compiling/testing, but I don't like working like that.

--- a/docs/docs/testing.md
+++ b/docs/docs/testing.md
@@ -43,7 +43,7 @@ describe("ParseService", () => {
 });
 ```
 
-Testing asynchronous method is also possible with `Done` function:
+Testing asynchronous method is also possible using `Promise`s (`async`/`await`):
 
 ```typescript
 import {expect} from "chai";
@@ -52,20 +52,15 @@ import {DbService} from "../services/db";
 
 describe("DbService", () => {
     let result: any;
-    before(inject([DbService, Done], (dbService: DbService, done: Done) => {
-        dbService
-        .getData()
-        .then((data) => {
-            result = data;
-            done();
-        });
+    before(inject([DbService, Done], async (dbService: DbService) => {
+        result = await dbService.getData();
     }));
     it("should data from db", () => {
         expect(result).to.be.an("object");
     });
 });
 ```
-> You can use also the `chai-promised` librairy for asynchronous test base on promise.
+> You can use also the `chai-promised` library for asynchronous tests based on promise.
 
 ### Testing controllers
 #### basic usage
@@ -74,7 +69,7 @@ Use `InjectorService` to get your controller from injector and test it:
 
 ```typescript
 import {expect} from "chai";
-import {inject, bootstrap} from "@tsed/testing";
+import {inject, bootstrap, TestContext} from "@tsed/testing";
 import {MyCtrl} from "../controllers/MyCtrl";
 import {Server} from "../Server";
 
@@ -87,6 +82,8 @@ describe("MyCtrl", () => {
        instance = calendarCtrl
     }))
 
+    after(TestContext.reset); // Make sure to clean up the global injector afterwards
+
     it("should do something", () => {
         expect(!!myCtrl).to.be.true;
     });
@@ -98,7 +95,7 @@ Or invoke a new instance of your controller like this:
 ```typescript
 import {expect} from "chai";
 import {InjectorService} from "@tsed/common";
-import {inject, bootstrap} from "@tsed/testing";
+import {inject, bootstrap, TestContext} from "@tsed/testing";
 import {MyCtrl} from "../controllers/MyCtrl";
 
 describe("MyCtrl", () => {
@@ -109,6 +106,8 @@ describe("MyCtrl", () => {
     before(inject([InjectorService], (injectorService: InjectorService) => {
        instance = InjectorService.invoke(MyCtrl);
     }))
+
+    after(TestContext.reset);
 
     it("should do something", () => {
         expect(!!instance).to.be.true;
@@ -139,7 +138,7 @@ export class MyCtrl {
 ```typescript
 // in MyCtrl.spec.ts
 import {expect} from "chai";
-import {inject} from "@tsed/testing";
+import {inject, TestContext} from "@tsed/testing";
 import {MyCtrl} from "../controllers/MyCtrl";
 import {DbService} from "../services/DbService";
 
@@ -147,6 +146,7 @@ describe("MyCtrl", () => {
 
     // bootstrap your Server to load all endpoints before run your test
     before(bootstrap(Server));
+    after(TestContext.reset);
 
     it("should do something", inject([InjectorService], (injector: InjectorService) => {
         
@@ -286,7 +286,7 @@ npm install --save-dev supertest @types/supertest
 
 ```typescript
 import {ExpressApplication} from "@tsed/common";
-import {bootstrap, inject} from "@tsed/testing";
+import {bootstrap, inject, TestContext} from "@tsed/testing";
 import * as SuperTest from "supertest";
 import {expect} from "chai";
 import {Server} from "../Server";
@@ -294,6 +294,7 @@ import {Server} from "../Server";
 describe("Rest", () => {
     // bootstrap your Server to load all endpoints before run your test
     beforeEach(bootstrap(Server));
+    afterEach(TestContext.reset);
 
     describe("GET /rest/calendars", () => {
         let app;

--- a/packages/testing/src/TestContext.ts
+++ b/packages/testing/src/TestContext.ts
@@ -1,0 +1,11 @@
+import {InjectorService} from "@tsed/di";
+
+export class TestContext {
+  static injector: InjectorService | null;
+  static reset() {
+    if (this.injector) {
+      this.injector.clear();
+    }
+    this.injector = null;
+  }
+}

--- a/packages/testing/src/TestContext.ts
+++ b/packages/testing/src/TestContext.ts
@@ -1,11 +1,15 @@
 import {InjectorService} from "@tsed/di";
 
 export class TestContext {
-  static injector: InjectorService | null;
+  static injector: InjectorService | null = null;
+
+  /**
+   * Resets the test injector of the test context, so it won't pollute your next test. Call this in your `tearDown` logic.
+   */
   static reset() {
-    if (this.injector) {
-      this.injector.clear();
+    if (TestContext.injector && TestContext.injector.clear) {
+      TestContext.injector.clear();
     }
-    this.injector = null;
+    TestContext.injector = null;
   }
 }

--- a/packages/testing/src/bootstrap.ts
+++ b/packages/testing/src/bootstrap.ts
@@ -1,18 +1,17 @@
 import {$log} from "ts-log-debug";
-import {TestContext} from "./testContext";
+import {TestContext} from "./TestContext";
 
 /**
  * Load the server silently without listening port and configure it on test profile.
  * @decorator
  * @param server
  * @param args
- * @returns {(done:Function)=>undefined}
+ * @returns {Promise<void>}
  */
-export function bootstrap(server: any, ...args: any[]) {
+export function bootstrap(server: any, ...args: any[]): () => Promise<void> {
   $log.stop();
-  process.env.NODE_ENV = process.env.NODE_ENV || "test";
 
-  return function before(done: any) {
+  return function before() {
     const instance = new server(...args);
 
     instance.startServers = () => Promise.resolve();
@@ -20,30 +19,6 @@ export function bootstrap(server: any, ...args: any[]) {
     // used by inject method
     TestContext.injector = instance.injector;
 
-    instance
-      .start()
-      .then(done)
-      .catch(done);
+    return instance.start();
   };
-}
-
-/**
- * Load the server silently without listening port and configure it on test profile.
- * Same as `bootstrap`, but this returns a promise instead
- * @example await bootstrapAsPromised(server);
- * @decorator
- * @param server
- * @param args
- * @returns {(done:Function)=>undefined}
- */
-export function bootstrapAsPromised(server: any, ...args: any[]): Promise<void> {
-  return new Promise((res, rej) => {
-    bootstrap(server, ...args)((err: Error) => {
-      if (err) {
-        rej(err);
-      } else {
-        res();
-      }
-    });
-  });
 }

--- a/packages/testing/src/bootstrap.ts
+++ b/packages/testing/src/bootstrap.ts
@@ -1,6 +1,5 @@
 import {$log} from "ts-log-debug";
-
-// export const servers: Map<any, any> = new Map();
+import {TestContext} from "./testContext";
 
 /**
  * Load the server silently without listening port and configure it on test profile.
@@ -18,12 +17,33 @@ export function bootstrap(server: any, ...args: any[]) {
 
     instance.startServers = () => Promise.resolve();
 
-    // TODO used by inject method
-    this.$$injector = instance.injector;
+    // used by inject method
+    TestContext.injector = instance.injector;
 
     instance
       .start()
       .then(done)
       .catch(done);
   };
+}
+
+/**
+ * Load the server silently without listening port and configure it on test profile.
+ * Same as `bootstrap`, but this returns a promise instead
+ * @example await bootstrapAsPromised(server);
+ * @decorator
+ * @param server
+ * @param args
+ * @returns {(done:Function)=>undefined}
+ */
+export function bootstrapAsPromised(server: any, ...args: any[]): Promise<void> {
+  return new Promise((res, rej) => {
+    bootstrap(server, ...args)((err: Error) => {
+      if (err) {
+        rej(err);
+      } else {
+        res();
+      }
+    });
+  });
 }

--- a/packages/testing/src/done.ts
+++ b/packages/testing/src/done.ts
@@ -1,4 +1,0 @@
-export type Done = Function;
-
-// tslint:disable-next-line: variable-name
-export const Done = Symbol("mocha:done");

--- a/packages/testing/src/index.ts
+++ b/packages/testing/src/index.ts
@@ -1,5 +1,4 @@
 export * from "./inject";
-export * from "./done";
 export * from "./bootstrap";
 export * from "./invoke";
 export * from "./loadInjector";

--- a/packages/testing/src/index.ts
+++ b/packages/testing/src/index.ts
@@ -3,3 +3,4 @@ export * from "./done";
 export * from "./bootstrap";
 export * from "./invoke";
 export * from "./loadInjector";
+export * from "./TestContext";

--- a/packages/testing/src/inject.ts
+++ b/packages/testing/src/inject.ts
@@ -1,5 +1,6 @@
 import {Done} from "./done";
 import {loadInjector} from "./loadInjector";
+import {TestContext} from "./testContext";
 
 /**
  * The inject function is one of the TsExpressDecorator testing utilities.
@@ -16,7 +17,7 @@ import {loadInjector} from "./loadInjector";
  */
 export function inject(targets: any[], func: Function) {
   return function before(done: Function) {
-    const injector = this.$$injector || loadInjector();
+    const injector = TestContext.injector || loadInjector();
 
     let isDoneInjected = false;
     const args = targets.map(target => {

--- a/packages/testing/src/inject.ts
+++ b/packages/testing/src/inject.ts
@@ -1,4 +1,3 @@
-import {Done} from "./done";
 import {loadInjector} from "./loadInjector";
 import {TestContext} from "./TestContext";
 
@@ -10,34 +9,13 @@ import {TestContext} from "./TestContext";
  *
  * * an array of Service dependency injection tokens,
  * * a test function whose parameters correspond exactly to each item in the injection token array.
- *
- * @param targets
- * @param func
- * @returns {any}
  */
-export function inject(targets: any[], func: Function) {
-  return function before(done: Function) {
+export function inject<T>(targets: any[], func: (...args: any[]) => Promise<T> | T): () => Promise<T> {
+  return async (): Promise<T> => {
     const injector = TestContext.injector || loadInjector();
 
-    let isDoneInjected = false;
-    const args = targets.map(target => {
-      if (target === Done) {
-        isDoneInjected = true;
-
-        return done;
-      }
-
-      /* istanbul ignore next */
-      if (!injector.has(target)) {
-        return injector.invoke(target);
-      }
-
-      return injector.get(target);
-    });
-
-    const result = func.apply(null, args);
-
-    if (!isDoneInjected) done();
+    const args = targets.map(target => (injector.has(target) ? injector.get(target) : injector.invoke(target)));
+    const result = await func(...args);
 
     return result;
   };

--- a/packages/testing/src/inject.ts
+++ b/packages/testing/src/inject.ts
@@ -1,6 +1,6 @@
 import {Done} from "./done";
 import {loadInjector} from "./loadInjector";
-import {TestContext} from "./testContext";
+import {TestContext} from "./TestContext";
 
 /**
  * The inject function is one of the TsExpressDecorator testing utilities.

--- a/test/integration/di.spec.ts
+++ b/test/integration/di.spec.ts
@@ -1,5 +1,5 @@
 import {InjectorService, ProviderScope} from "@tsed/common";
-import {bootstrap, inject} from "@tsed/testing";
+import {bootstrap, inject, TestContext} from "@tsed/testing";
 import {expect} from "../tools";
 import {ProductsCtrl} from "./app/controllers/products/ProductsCtrl";
 import {InnerService} from "./app/services/InnerService";
@@ -7,9 +7,9 @@ import {OuterService} from "./app/services/OuterService";
 import {FakeServer} from "./FakeServer";
 
 describe("DI", () => {
-  before(bootstrap(FakeServer));
-  before(
-    inject([InjectorService], (injector: InjectorService) => {
+  before(async () => {
+    await bootstrap(FakeServer)();
+    await inject([InjectorService], (injector: InjectorService) => {
       this.locals = new Map<string | Function, any>();
       const provider = injector.getProvider(ProductsCtrl)!;
       const target = provider.useClass;
@@ -17,8 +17,10 @@ describe("DI", () => {
       this.rebuildHandler = provider.scope !== ProviderScope.SINGLETON;
 
       this.instance = injector.invoke(target, this.locals, undefined, true);
-    })
-  );
+    })();
+  });
+
+  after(TestContext.reset);
 
   it("should setting rebuild handler to true", () => {
     expect(this.rebuildHandler).to.eq(true);

--- a/test/integration/rest.spec.ts
+++ b/test/integration/rest.spec.ts
@@ -1,5 +1,5 @@
 import {ExpressApplication} from "@tsed/common";
-import {bootstrap, inject} from "@tsed/testing";
+import {bootstrap, inject, TestContext} from "@tsed/testing";
 import * as SuperTest from "supertest";
 import {expect} from "../tools";
 import {FakeServer} from "./FakeServer";
@@ -11,6 +11,7 @@ describe("Rest", () => {
       this.app = SuperTest(expressApplication);
     })
   );
+  after(TestContext.reset);
   describe("integration", () => {
     describe("GET /rest", () => {
       it("should return html content", done => {

--- a/test/integration/swagger.spec.ts
+++ b/test/integration/swagger.spec.ts
@@ -1,5 +1,5 @@
 import {ExpressApplication} from "@tsed/common";
-import {bootstrap, inject} from "@tsed/testing";
+import {bootstrap, inject, TestContext} from "@tsed/testing";
 import * as SuperTest from "supertest";
 import {expect} from "../tools";
 import {FakeServer} from "./FakeServer";
@@ -7,6 +7,7 @@ import {FakeServer} from "./FakeServer";
 describe("Swagger", () => {
   before(bootstrap(FakeServer));
   before(inject([ExpressApplication], (expressApplication: ExpressApplication) => (this.app = SuperTest(expressApplication))));
+  after(TestContext.reset);
 
   describe("GET /api-doc/swagger.json", () => {
     before(done => {

--- a/test/integration/testing-example.spec.ts
+++ b/test/integration/testing-example.spec.ts
@@ -1,5 +1,5 @@
 import {Controller, Get, InjectorService, ParseService, Service} from "@tsed/common";
-import {bootstrap, inject} from "@tsed/testing";
+import {bootstrap, inject, TestContext} from "@tsed/testing";
 import {expect} from "chai";
 import {AcceptMimesMiddleware} from "../../packages/common/src/mvc/components/AcceptMimesMiddleware";
 import {Hidden} from "../../packages/swagger/src";
@@ -66,6 +66,7 @@ describe("Example Test", () => {
         instance = calendarCtrl;
       })
     );
+    after(TestContext.reset);
 
     it("should do something", () => {
       expect(!!instance).to.be.true;
@@ -82,6 +83,7 @@ describe("Example Test", () => {
         instance = injectorService.invoke(CalendarCtrl);
       })
     );
+    after(TestContext.reset);
 
     it("should do something", () => {
       expect(!!instance).to.be.true;
@@ -91,6 +93,7 @@ describe("Example Test", () => {
   describe("Mock dependencies", () => {
     // bootstrap your Server to load all endpoints before run your test
     before(bootstrap(FakeServer));
+    after(TestContext.reset);
 
     it("should do something", inject([InjectorService], (injector: InjectorService) => {
       // create locals map

--- a/test/units/filters/services/ParseService.spec.ts
+++ b/test/units/filters/services/ParseService.spec.ts
@@ -1,6 +1,6 @@
 import {expect} from "chai";
 import {ParseService} from "../../../../packages/common/src";
-import {Done, inject} from "../../../../packages/testing/src";
+import {inject} from "../../../../packages/testing/src";
 
 describe("ParseService :", () => {
   it("should clone object", () => {
@@ -9,16 +9,14 @@ describe("ParseService :", () => {
     expect(ParseService.clone(source)).not.to.be.equal(source);
   });
 
-  it("should not clone object", inject([ParseService, Done], (parserService: ParseService, done: Function) => {
+  it("should not clone object", inject([ParseService], (parserService: ParseService) => {
     const source = {test: {}};
 
     expect(parserService.eval(undefined!, source, false)).to.equal(source);
     expect(parserService.eval("test", source, false)).to.equal(source.test);
-
-    done();
   }));
 
-  it("should eval expression with a scope and return value", inject([ParseService, Done], (parserService: ParseService, done: Function) => {
+  it("should eval expression with a scope and return value", inject([ParseService], (parserService: ParseService) => {
     expect(
       parserService.eval(undefined!, {
         test: "yes"
@@ -26,18 +24,14 @@ describe("ParseService :", () => {
     ).to.equal("yes");
 
     expect(parserService.eval(undefined!, "test")).to.equal("test");
-
-    done();
   }));
 
-  it("should eval expression with a scope and return value", inject([ParseService, Done], (parserService: ParseService, done: Function) => {
+  it("should eval expression with a scope and return value", inject([ParseService], (parserService: ParseService) => {
     expect(
       parserService.eval("test", {
         test: "yes"
       })
     ).to.equal("yes");
-
-    done();
   }));
 
   it("should eval expression with a scope and return value", inject([ParseService], (parserService: ParseService) => {

--- a/test/units/testing/TestContext.spec.ts
+++ b/test/units/testing/TestContext.spec.ts
@@ -1,0 +1,17 @@
+import {loadInjector, TestContext} from "@tsed/testing";
+import {Provider} from "@tsed/di";
+import {expect} from "../../tools";
+
+describe("TestContext", () => {
+  describe("reset()", () => {
+    it("should reset the injector", () => {
+      const injector = loadInjector();
+      const injectionKey = "key";
+      TestContext.injector = injector;
+      TestContext.injector.set(injectionKey, new Provider("something"));
+      TestContext.reset();
+      expect(injector).lengthOf(0);
+      expect(TestContext.injector).eq(null);
+    });
+  });
+});

--- a/test/units/testing/bootstrap.spec.ts
+++ b/test/units/testing/bootstrap.spec.ts
@@ -1,0 +1,39 @@
+import {bootstrap, TestContext} from "@tsed/testing";
+import {expect} from "../../tools";
+
+class FakeServer {
+  static current: FakeServer;
+  startServers: Function;
+
+  injector = "Injector";
+
+  async start() {
+    FakeServer.current = this;
+
+    return this.startServers();
+  }
+}
+
+describe("bootstrap()", () => {
+  const fn = bootstrap(FakeServer);
+
+  beforeEach(fn);
+  afterEach(TestContext.reset);
+
+  it("should return before function", () => {
+    expect(fn).to.be.a("function");
+  });
+
+  it("should return promise after executing the function", () => {
+    expect(fn()).instanceOf(Promise);
+  });
+
+  it("should attach injector instance to TestContext", () => {
+    expect(TestContext.injector).to.eq("Injector");
+  });
+
+  it("should replace FakeServer.startServers by a stub()", () => {
+    expect(FakeServer.current.startServers).to.be.a("Function");
+    expect(FakeServer.current.startServers()).to.be.an.instanceOf(Promise);
+  });
+});

